pkg/commands/products/root.go

```diff
@@ -0,0 +1,189 @@
+// It should be installed under the primary root command.
+type RootCommand struct {
+	cmd.Base
+	cmd.JSONOutput
+
+	disableProduct string
+	enableProduct  string
+	manifest       manifest.Data
+	serviceName    cmd.OptionalServiceNameID
+}
+
+// ProductEnablementOptions is a list of products that can be enabled/disabled.
+var ProductEnablementOptions = []string{
+	"brotli_compression",
+	"domain_inspector",
+	"fanout",
+	"image_optimizer",
+	"origin_inspector",
+	"websockets",
+}
+
+// NewRootCommand returns a new command registered in the parent.
+func NewRootCommand(parent cmd.Registerer, g *global.Data, m manifest.Data) *RootCommand {
+	var c RootCommand
+	c.Globals = g
+	c.manifest = m
+	c.CmdClause = parent.Command("products", "Enable, disable, and check the enablement status of products on your services")
+
+	// Optional.
+	c.CmdClause.Flag("disable", "Disable product").HintOptions(ProductEnablementOptions...).EnumVar(&c.disableProduct, ProductEnablementOptions...)
+	c.CmdClause.Flag("enable", "Enable product").HintOptions(ProductEnablementOptions...).EnumVar(&c.enableProduct, ProductEnablementOptions...)
+	c.RegisterFlagBool(c.JSONFlag()) // --json
+	c.RegisterFlag(cmd.StringFlagOpts{
+		Name:        cmd.FlagServiceIDName,
+		Description: cmd.FlagServiceIDDesc,
+		Dst:         &c.manifest.Flag.ServiceID,
+		Short:       's',
+	})
+	c.RegisterFlag(cmd.StringFlagOpts{
+		Action:      c.serviceName.Set,
+		Name:        cmd.FlagServiceName,
+		Description: cmd.FlagServiceDesc,
+		Dst:         &c.serviceName.Value,
+	})
+	return &c
+}
+
+// Exec implements the command interface.
+func (c *RootCommand) Exec(_ io.Reader, out io.Writer) error {
+	ac := c.Globals.APIClient
+
+	if c.enableProduct != "" && c.disableProduct != "" {
+		return fsterr.ErrInvalidProductEnablementFlagCombo
+	}
+
+	if c.Globals.Verbose() && c.JSONOutput.Enabled {
+		return fsterr.ErrInvalidVerboseJSONCombo
+	}
+
+	serviceID, _, _, err := cmd.ServiceID(c.serviceName, c.manifest, c.Globals.APIClient, c.Globals.ErrLog)
+	if err != nil {
+		return fmt.Errorf("failed to identify Service ID: %w", err)
+	}
+
+	if c.enableProduct != "" {
+		if _, err := ac.EnableProduct(&fastly.ProductEnablementInput{
+			ProductID: identifyProduct(c.enableProduct),
+			ServiceID: serviceID,
+		}); err != nil {
+			return fmt.Errorf("failed to enable product '%s': %w", c.enableProduct, err)
+		}
+		text.Success(out, "Successfully enabled product '%s'", c.enableProduct)
+		return nil
+	}
+
+	if c.disableProduct != "" {
+		if err := ac.DisableProduct(&fastly.ProductEnablementInput{
+			ProductID: identifyProduct(c.disableProduct),
+			ServiceID: serviceID,
+		}); err != nil {
+			return fmt.Errorf("failed to disable product '%s': %w", c.disableProduct, err)
+		}
+		text.Success(out, "Successfully disabled product '%s'", c.disableProduct)
+		return nil
+	}
+
+	var brotliEnabled, diEnabled, fanoutEnabled, ioEnabled, oiEnabled, wsEnabled bool
```

Comment:
nitpick: These variables could be replaced by an inline struct. It could even be the same struct as used to generate the JSON output.

Reply:
Done. Looks much better thanks.